### PR TITLE
New normalization for fourier lattices

### PR DESCRIPTION
**Pull request by AliGhorashiCMT**

### CODE REVIEW

**Comment by thchr** [comment]
Can we merge this method with the method above and let the `Gs` argument be the third argument of `normscale!` with union type `Union{Nothing, ReciprocalBasis{D}}` and default value `nothing`? Then you can just branch on the type of `Gs` when you calculate the rescaling factor.

(And then we could also clean up/remove/reduce the comment in the old `normscale!` that warns about this not being a great way to do things)

**Comment by thchr** [comment]
Can we merge this method with the method above and let the `Gs` argument be the third argument of `normscale!` with union type `Union{Nothing, ReciprocalBasis{D}}` and default value `nothing`? Then you can just branch on the type of `Gs` when you calculate the rescaling factor.

(And then we could also clean up/remove/reduce the comment in the old `normscale!` that warns about this not being a great way to do things)

**Comment by thchr** [comment]
(and thanks!)

**Comment by thchr** [comment]
(and thanks!)

**Comment by AliGhorashiCMT** [comment]
Hi! I just streamlined it a bit as per your suggestions

File: src/lattices.jl [hunk 1]
```diff
@@ -377,37 +377,21 @@ normscale(flat::ModulatedFourierLattice, expon::Real) = normscale!(deepcopy(flat
...
+                rescale_factor = norm(first(orbits[i]))^expon
+                rescale_factor == zero(rescale_factor) && continue # for G = [0,0,0] case
+                flat.orbitcoefs[i] ./= rescale_factor
+            end
+        else
+            @inbounds for i in eachindex(orbits)
+                rescale_factor = norm(dot(Gs, first(orbits[i])))^expon
+                rescale_factor == zero(rescale_factor) && continue # for G = [0,0,0] case
+                flat.orbitcoefs[i] ./= rescale_factor
+            end
```
Comment (thchr) [inline src/lattices.jl:394]: You could avoid some code-duplication by branching inside the loop (the branch is resolved statically, so there's no performance downside):
```jl
      @inbounds for i in eachindex(orbits)
         n = if isnothing(Gs)
            norm(first(orbits[i]))
         else
            norm(dot(Gs, first(orbits[i])))
         end
         rescale_factor = n^expon
         rescale_factor == zero(rescale_factor) && continue # for G = [0,0,0] case
         flat.orbitcoefs[i] ./= rescale_factor
      end
  ```

File: src/lattices.jl [hunk 2]
```diff
@@ -377,37 +377,21 @@ normscale(flat::ModulatedFourierLattice, expon::Real) = normscale!(deepcopy(flat
 
 In-place equivalent of `normscale`: changes `flat`.
 """
-function normscale!(flat::ModulatedFourierLattice, expon::Real)
-    if !iszero(expon)
+function normscale!(flat::ModulatedFourierLattice, Gs::Union{ReciprocalBasis{D}, Nothing} = nothing, expon::Real) where D
```
Comment (thchr) [inline src/lattices.jl]: The `Gs` argument has to be the last argument in order to have a default value:
```suggestion
function normscale!(flat::ModulatedFourierLattice,
                    expon::Real,
                    Gs::Union{ReciprocalBasis{D}, Nothing} = nothing) where D
```

**Comment by thchr** [comment]
Nice, thanks. Aside from the two review comments I left just now, can you also update the docstrings of `normscale` and `normscale!`? Thanks!

File: src/lattices.jl (hunk 2 above)
Comment (thchr) [reply src/lattices.jl]: I see you did this before I actually managed to get the comments "published" (always find the commenting interface fiddly...)

**Comment by AliGhorashiCMT** [comment]
Hi- yeah I changed the Gs to be the final value a commit ago I think. I changed the docstrings just now.

File: src/lattices.jl [hunk 3]
```diff
@@ -366,41 +366,38 @@ function modulate(flat::AbstractFourierLattice{D},
...
-    normscale(flat::ModulatedFourierLattice, expon::Real) --> ModulatedFourierLattice
+    normscale(flat::ModulatedFourierLattice, expon::Real, Gs::Union{ReciprocalBasis{D}, Nothing} = nothing) where D --> ModulatedFourierLattice
 
-Applies subsequent norm-rescaling via `expon`; see detailed description 
-in `modulate`. An in-place variant is provided as `normscale!`.
+Applies subsequent norm-rescaling via `expon`; Passing in `Gs` as a ReciprocalBasis enforces norm scaling in a Cartesian basis- otherwise norms are computed in the reciprocal lattice basis.
+see detailed description in `modulate`. An in-place variant is provided as `normscale!`.
 """
-normscale(flat::ModulatedFourierLattice, expon::Real) = normscale!(deepcopy(flat), expon)
+normscale(flat::ModulatedFourierLattice, expon::Real, Gs::Union{ReciprocalBasis{D}, Nothing} = nothing)  where D = normscale!(deepcopy(flat), expon, Gs)
```
Comment (thchr) [inline src/lattices.jl:374]:
```suggestion
normscale(flat::ModulatedFourierLattice{D}, expon::Real, Gs::Union{ReciprocalBasis{D}, Nothing} = nothing)  where D = normscale!(deepcopy(flat), expon, Gs)
```

File: src/lattices.jl [hunk 4]
```diff
@@ -366,41 +366,38 @@ function modulate(flat::AbstractFourierLattice{D},
 end
 
 @doc """ 
-    normscale(flat::ModulatedFourierLattice, expon::Real) --> ModulatedFourierLattice
+    normscale(flat::ModulatedFourierLattice, expon::Real, Gs::Union{ReciprocalBasis{D}, Nothing} = nothing) where D --> ModulatedFourierLattice
```
Comment (thchr) [inline src/lattices.jl:369]:
```suggestion
    normscale(flat::ModulatedFourierLattice, expon::Real, Gs::Union{ReciprocalBasis, Nothing} = nothing)  --> ModulatedFourierLattice
```

File: src/lattices.jl [hunk 5]
```diff
@@ -366,41 +366,38 @@ function modulate(flat::AbstractFourierLattice{D},
...
-Applies subsequent norm-rescaling via `expon`; see detailed description 
-in `modulate`. An in-place variant is provided as `normscale!`.
+Applies subsequent norm-rescaling via `expon`; Passing in `Gs` as a ReciprocalBasis enforces norm scaling in a Cartesian basis- otherwise norms are computed in the reciprocal lattice basis.
+see detailed description in `modulate`. An in-place variant is provided as `normscale!`.
 """
-normscale(flat::ModulatedFourierLattice, expon::Real) = normscale!(deepcopy(flat), expon)
+normscale(flat::ModulatedFourierLattice, expon::Real, Gs::Union{ReciprocalBasis{D}, Nothing} = nothing)  where D = normscale!(deepcopy(flat), expon, Gs)
 @doc """
-    normscale!(flat::ModulatedFourierLattice, expon::Real) --> ModulatedFourierLattice
+    normscale!(flat::ModulatedFourierLattice, expon::Real, Gs::Union{ReciprocalBasis{D}, Nothing} = nothing) where D --> ModulatedFourierLattice
```
Comment (thchr) [inline src/lattices.jl:376]:
```suggestion
    normscale!(flat::ModulatedFourierLattice, expon::Real, Gs::Union{ReciprocalBasis, Nothing} = nothing) --> ModulatedFourierLattice
```

File: src/lattices.jl [hunk 6]
```diff
@@ -366,41 +366,38 @@ function modulate(flat::AbstractFourierLattice{D},
...
 
 In-place equivalent of `normscale`: changes `flat`.
 """
-function normscale!(flat::ModulatedFourierLattice, expon::Real)
-    if !iszero(expon)
+function normscale!(flat::ModulatedFourierLattice,  expon::Real, Gs::Union{ReciprocalBasis{D}, Nothing} = nothing) where D
+    if !iszero(expon)
```
Comment (thchr) [inline src/lattices.jl:381]: Whitespace:
```suggestion
    if !iszero(expon)
```

File: src/lattices.jl [hunk 7]
```diff
@@ -366,41 +366,38 @@ function modulate(flat::AbstractFourierLattice{D},
...
+normscale(flat::ModulatedFourierLattice, expon::Real, Gs::Union{ReciprocalBasis{D}, Nothing} = nothing)  where D = normscale!(deepcopy(flat), expon, Gs)
 @doc """
-    normscale!(flat::ModulatedFourierLattice, expon::Real) --> ModulatedFourierLattice
+    normscale!(flat::ModulatedFourierLattice, expon::Real, Gs::Union{ReciprocalBasis{D}, Nothing} = nothing) where D --> ModulatedFourierLattice
 
 In-place equivalent of `normscale`: changes `flat`.
 """
-function normscale!(flat::ModulatedFourierLattice, expon::Real)
-    if !iszero(expon)
+function normscale!(flat::ModulatedFourierLattice,  expon::Real, Gs::Union{ReciprocalBasis{D}, Nothing} = nothing) where D
```
Comment (thchr) [inline src/lattices.jl:380]:
```suggestion
function normscale!(flat::ModulatedFourierLattice{D}, expon::Real, Gs::Union{ReciprocalBasis{D}, Nothing} = nothing) where D
```

File: src/lattices.jl [hunk 8]
```diff
@@ -366,41 +366,38 @@ function modulate(flat::AbstractFourierLattice{D},
...
+                rescale_factor = norm(dot(Gs, first(orbits[i])))^expon
+                rescale_factor == zero(rescale_factor) && continue # for G = [0,0,0] case
+                flat.orbitcoefs[i] ./= rescale_factor
+            end
         end
     end
     return flat
 end
 
+
```
Comment (thchr) [inline src/lattices.jl:400]:
```suggestion
```

File: src/lattices.jl [hunk 9]
```diff
@@ -366,41 +366,38 @@ function modulate(flat::AbstractFourierLattice{D},
 end
 
 @doc """ 
-    normscale(flat::ModulatedFourierLattice, expon::Real) --> ModulatedFourierLattice
+    normscale(flat::ModulatedFourierLattice, expon::Real, Gs::Union{ReciprocalBasis{D}, Nothing} = nothing) where D --> ModulatedFourierLattice
 
-Applies subsequent norm-rescaling via `expon`; see detailed description 
-in `modulate`. An in-place variant is provided as `normscale!`.
+Applies subsequent norm-rescaling via `expon`; Passing in `Gs` as a ReciprocalBasis enforces norm scaling in a Cartesian basis- otherwise norms are computed in the reciprocal lattice basis.
```
Comment (thchr) [inline src/lattices.jl:371]: Generally try to keep lines shorter than 80 characters if possible.

**Comment by AliGhorashiCMT** [comment]
I think I committed the suggestions- should we also change the modulate function so it has the same normalization scheme?

File: src/lattices.jl (hunk 1 above)
Comment (thchr) [reply src/lattices.jl:394]: Would be good to fix this before merging also.

**Comment by thchr** [comment]
Thanks for the edits; I'll merge now!

File: src/lattices.jl [hunk 10]
```diff
@@ -355,7 +357,12 @@ function modulate(flat::AbstractFourierLattice{D},
     if !isnothing(expon) && !iszero(expon) 
         @inbounds for i in 2:length(orbits) # leaves the constant term untouched 
                                             # (there will _always_ be a constant term)...
-            modulation[i] /= (norm(first(orbits[i])))^expon
+            n = if isnothing(Gs)
+                norm(first(orbits[i]))^expon
```
Comment (thchr) [inline src/lattices.jl:361]: Ouch; we doubled up on `^expon` here. Fixed in https://github.com/thchr/Crystalline.jl/commit/387362ae5d5c2f24be36cf83e5897ad2ff1cb56c :)

**Pull request merged**